### DISCUSSION TITLE
ChatGPT created code

### DISCUSSION THREAD

**Discussion by mmliam** [micropython/micropython#11755] [General]
When asked: "Create a program in MicroPython for an ESP8266 to loop receiving an ESPNow message from an ESP32 access point, then print the result"
This is the code it created:
```
import network
import espnow

wlan = network.WLAN(network.STA_IF)
wlan.active(True)

esp = espnow.ESPNow()
esp.init()

def on_recv(p):
    print(p)

esp.on_recv(on_recv)

while True:
    esp.check_msg()
```

Does this make any sense to anyone?

I had made the exact same request after clearing all EDGE cookies.
This was the result:
```
import network
import espnow

wlan = network.WLAN(network.STA_IF)
wlan.active(True)

espnow.init()
espnow.add_peer(b'\x30\xae\xa4\x1f\x5d\x2c', ifidx=espnow.ESP_IF_WIFI_STA)

while True:
    msg = espnow.recv_msg()
    if msg is not None:
        print(msg)

```
There is a "New Topic" icon that will probably clear ChatGPT topic memory, as deleting the cookies did.
Go figure; it got it right; at least close. ;-)

**Comment by 2dof** [comment]
ChatGPT  is generative LLM, do not exactly fit for  creating code . On this level of 'evolution' it just will search www to find "answer" for your question, then it is make sence (i.e find answer for question in linguistic meaning) but from expecting a "write/generate" code it is Russian roulette ( for simple question 1 bullet, for more complicated: 6 in 6 chamber).  
In Your problem it make  sense because chatGBT found code in www (work as search engine) ,   and can find code like in :
 https://docs.micropython.org/en/latest/library/espnow.html.
or 
https://docs.petoi.com/apis/8266-micropython-controller/using-the-esp-now-protocol

When you as question in google search, then you will got list of www (github, tutotrials, etc) and then you will read, got some links to source, etc, You will understand context and will be able to chose best fitted solution for your problem, most important you will learn something about programming, hardware etc..  Chat GBT will do not perform this process - because it looking answer from linguistic point of view (not programmer perspective). 

My suggestion: do not use GBT for programming, unless it itself  is some experiment

**Reply by mmliam** [reply]
I was just fooling around with it; testing it to see the quality of the code generated. When I initially tried, it gave reasonable results. But, the more I tried with slight re-wording of the request, first it got better. Then it returned with the first sample I posted, which appears to be nonsense, and no matter how I changed the wording it stuck with that same output.  Then I cleared the browser cookies, and it returned a reasonable response again.

It seems the more you try with similar wording ChatGPT assumes you don't like the last response and tries something different. It also keeps a memory of past responses, referenced in the cookies.

**Reply by 2dof** [reply]
chatGBT has some programming languages information rules incorporated but more use like to mimic human debugging (but it already exist in IDE's). Sooner or later it will be able to "generate" code but evolution will be first in direction of  web development programming languages and frameworks. There were test wit C++ but it ended with refactoring by human after all.  
Your observations and conclusions about behaviour of chatGBT shows that it is not  heuristic an more just "template matching" system, and it is enough (for working) , but there is still not enough learning data ( code sources) for micro python.

**Comment by jimmo** [comment]
> Does this make any sense to anyone?

Please refer to the official documentation for examples and API usage: https://docs.micropython.org/en/latest/library/espnow.html

**Comment by mmliam** [comment]
I have been testing ChatGPT as a reference source for Pytho/micropython instructions; so far I've tested the returned results and they have functioned correctly. What is really interesting is that you can ask follow-up questions about the returned results.

For a simple example I asked ChatGPT: Give an example of the Python bytes() function.

> Certainly! Here is an example of how to use the Python bytes() function:
> 
>  Example 1: Encoding a string into bytes
> string = "Hello World"
> bytes_string = bytes(string, 'utf-8')
> print(bytes_string)
> 
> Output: b'Hello World'
>
> In the above example, we have passed the string “Hello World” to the bytes() function with the encoding parameter as ‘utf-8’. The output is a bytes object which is prefixed with ‘b’.
> 
Next I asked: Why was the 'utf-8' encoding used instead of 'ascii'?

[Edited response, they can be lengthy]

> The reason why ‘utf-8’ encoding was used instead of ‘ascii’ is because ‘utf-8’ is capable of encoding all 1,112,064 valid character code points in Unicode using one to four one-byte code units.
> ASCII, on the other hand, can only encode 128 characters

Next I asked: Which encoding routine is faster; 'utf-8' or 'ascii'?

[Again lengthy response edited]

> ‘ascii’ encoding is faster than ‘utf-8’ encoding because ‘ascii’ is a single-byte encoding while ‘utf-8’ is a multi-byte encoding.

ChatGPT gives references, and justifications for all it's responses.

So far I'm quite impressed; as each response was returned in 10 secs; it would take me a lot longer to look-up references. Also, for the detailed information requested; differences between utf-8 & ascii encoding, particularly speed of operation might be very difficult, if not impossible to find in various references.
 While I am impressed with ChatGPT, it can go rogue, as I demonstrated in the beginning post. But, with some care, it can be a very useful tool that still learning and improving.

**Comment by dotnfc** [comment]
chatGPT can be a useful assistants, the better the question, the better the answer

**Reply by peterhinch** [reply]
I'm unconvinced. See [this thread](https://github.com/orgs/micropython/discussions/11895).
